promql/engine.go

```diff
@@ -700,16 +726,21 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 	enh := &EvalNodeHelper{out: make(Vector, 0, biggestLen)}
 	seriess := make(map[uint64]Series, biggestLen) // Output series by series hash.
 	for ts := ev.startTimestamp; ts <= ev.endTimestamp; ts += ev.interval {
+		// Reset num. of samples in memory after each timestamp.
+		tempNumSamples = ev.currentSamples
 		// Gather input vectors for this timestamp.
 		for i := range exprs {
 			vectors[i] = vectors[i][:0]
 			for si, series := range matrixes[i] {
 				for _, point := range series.Points {
-					if point.T == ts {
+					if point.T == ts && ev.currentSamples < ev.maxSamples {
 						vectors[i] = append(vectors[i], Sample{Metric: series.Metric, Point: point})
 						// Move input vectors forward so we don't have to re-scan the same
 						// past points at the next step.
 						matrixes[i][si].Points = series.Points[1:]
+						ev.currentSamples++
```

Comment:
@brian-brazil so now I assume that since we're adding the # of samples in the matrices returned by the calls to eval, we don't need to add to `ev.currentSamples` here, but we should keep the later block where we add the length of the `result` from `f`. Is that correct?

Reply:
No, we still need these as it is a (temporary) 2nd copy of some of the data in the input vector.